packages/client/src/components/app/CodeScanner.svelte

```diff
@@ -0,0 +1,235 @@
+          resolve({ initialised: false })
+        })
+    })
+  }
+
+  const checkCamera = async () => {
+    return new Promise(resolve => {
+      Html5Qrcode.getCameras()
+        .then(devices => {
+          if (devices && devices.length) {
+            resolve({ enabled: true })
+          }
+        })
+        .catch(e => {
+          console.error(e)
+          resolve({ enabled: false })
+        })
+    })
+  }
+
+  const start = async () => {
+    const status = await initReader()
+    cameraStarted = status.initialised
+  }
+
+  $: if (cameraEnabled && videoEle && !cameraStarted) {
+    start()
+  }
+
+  const showReaderModal = async () => {
+    camModal.show()
+    const camStatus = await checkCamera()
+    cameraEnabled = camStatus.enabled
+  }
+
+  const hideReaderModal = async () => {
+    cameraEnabled = undefined
+    cameraStarted = false
+    if (html5QrCode) {
+      await html5QrCode.stop()
+      html5QrCode = undefined
+    }
+    camModal.hide()
+  }
+</script>
+
+<div class="scanner-video-wrapper">
+  {#if value && !manualMode}
+    <div class="scanner-value field-display">
+      <StatusLight positive />
+      {value}
+    </div>
+  {/if}
+
+  {#if allowManualEntry && manualMode}
+    <div class="manual-input">
+      <Input
+        bind:value
+        on:change={() => {
+          dispatch("change", value)
+        }}
+      />
+    </div>
+  {/if}
+
+  {#if value}
+    <ActionButton
+      on:click={() => {
+        dispatch("change", "")
+      }}
+      {disabled}
+    >
+      Clear
+    </ActionButton>
+  {:else}
+    <ActionButton
+      icon="Camera"
+      on:click={() => {
+        showReaderModal()
+      }}
+      {disabled}
+    >
+      {scanButtonText}
+    </ActionButton>
+  {/if}
+</div>
+
+<div class="modal-wrap">
+  <Modal bind:this={camModal} on:hide={hideReaderModal}>
+    <ModalContent
+      title={scanButtonText}
+      showConfirmButton={false}
+      showCancelButton={false}
+    >
+      <div id="reader" class="container" bind:this={videoEle}>
+        <div class="camera-placeholder">
+          <Icon size="XXL" name="Camera" />
+          {#if cameraEnabled === false}
+            <div>Your camera is disabled.</div>
+          {/if}
+        </div>
+      </div>
+      {#if cameraEnabled === true}
+        <div class="code-wrap">
+          {#if value}
+            <div class="scanner-value">
+              <StatusLight positive />
+              {value}
+            </div>
+          {:else}
+            <div class="scanner-value">
+              <StatusLight neutral />
+              Searching for code...
+            </div>
+          {/if}
+        </div>
+      {/if}
+
+      <div slot="footer">
+        <div class="footer-buttons">
+          {#if allowManualEntry}
+            <Button
+              group
+              secondary
+              newStyles
+              on:click={() => {
+                manualMode = !manualMode
```

Comment:
NAB: since this is just a toggle, if you've enabled manual mode already and then press this again, it will unset it which feels a bit strange. I reckon we should probably just hide the button if we're already in manual mode, so `{#if allowManualEntry && !manualMode}`, and then the onclick handler here can just set it explicitly to `true`.